### ISSUE TITLE
sleep() ignored when there's a JS error

### ISSUE THREAD

**Issue by MaximizedAction** [loadimpact/k6#688]
The sleep() function is getting ignored when the JS code in the script produces an error during runtime.
Example:

**script.js:**
```
import {sleep} from 'k6';

export default function(){
        console.log("Iteration...");
        someNonsense
        sleep(3);
}
```
**Output:**
```
$ k6 run -i 5 script.js

          /\      |‾‾|  /‾‾/  /‾/
     /\  /  \     |  |_/  /  / /
    /  \/    \    |      |  /  ‾‾\
   /          \   |  |‾\  \ | (_) |
  / __________ \  |__|  \__\ \___/ .io

  execution: local
     output: -
     script: script.js

    duration: -,  iterations: 5
         vus: 1, max: 1

INFO[0001] Iteration...
ERRO[0001] ReferenceError: someNonesense is not defined
        at /home/user/script.js:5:2(7)

INFO[0001] Iteration...
ERRO[0001] ReferenceError: someNonesense is not defined
        at /home/user/script.js:5:2(7)

INFO[0001] Iteration...
ERRO[0001] ReferenceError: someNonesense is not defined
        at /home/user/script.js:5:2(7)

INFO[0001] Iteration...
ERRO[0001] ReferenceError: someNonesense is not defined
        at /home/user/script.js:5:2(7)

INFO[0001] Iteration...
ERRO[0001] ReferenceError: someNonesense is not defined
        at /home/user/script.js:5:2(7)

    done [==========================================================] 5 / 5

    iterations...: 5 4622.900741/s
    vus..........: 1 min=1 max=1
    vus_max......: 1 min=1 max=1
```

Compared to code that does not produce an error:
**script.js:**
```
import {sleep} from 'k6';

export default function(){
        console.log("Iteration...");
        //someNonsense
        sleep(3);
}
```
**Output:**
```
$ k6 run -i 5 script.js

          /\      |‾‾|  /‾‾/  /‾/
     /\  /  \     |  |_/  /  / /
    /  \/    \    |      |  /  ‾‾\
   /          \   |  |‾\  \ | (_) |
  / __________ \  |__|  \__\ \___/ .io

  execution: local
     output: -
     script: script.js

    duration: -,  iterations: 5
         vus: 1, max: 1

INFO[0001] Iteration...
INFO[0004] Iteration...
INFO[0007] Iteration...
INFO[0010] Iteration...
INFO[0013] Iteration...
    done [==========================================================] 5 / 5

    data_received........: 0 B 0 B/s
    data_sent............: 0 B 0 B/s
    iteration_duration...: avg=3s min=3s med=3s max=3s p(90)=3s p(95)=3s
    iterations...........: 5   0.333309/s
    vus..................: 1   min=1 max=1
    vus_max..............: 1   min=1 max=1
```

This is a trivial example, but the problem has been causing quite the hassle in more complex scripts. For example, when the test parameter isn't _iteration_, but _duration_. Judging by the iteration rate above, the output can spam 4623 error messages in 1 second.

**Comment by MaximizedAction** [comment]
Makes sense. And catching the error, or fixing it if possible, is the best way. Though I found it a bit unintuitive so to say, that the execution continues after an exception/error:

> just to be clear, sleep() is not ignored, the script execution just never reaches it because it's interrupted before that.

In retrospect, it's as you wrote, the script is just interrupted on every iteration. In my head the script corresponded to the test, so I expected the entire test to fail alltogether if it encountered an exception/error.

How about an option parameter that makes the test execution abort in case of a script exception/error?

**Comment by na--** [comment]
> How about an option parameter that makes the test execution abort in case of a script exception/error?

It's very likely that [goja](https://github.com/dop251/goja/), the JS runtime we use to execute the scripts, wouldn't support that, though I might be wrong - I'm not very familiar with it yet. And it would be a pretty strange and unexpected behavior to have in a language that has exceptions, so for sure the default value for such an option would have to be disabled. 

And if it's disabled by default, we're back to square one - if you know you can hit this issue with exceptions causing "instantaneous" iterations and want to guard against it, you can just use a global `try { /* your code */ } catch(e) {sleep(1); throw e; }`

**Comment by na--** [comment]
> How about an option parameter that makes the test execution abort in case of a script exception/error?

I think such an option has a lot of merit, but the tricky question is whether it should be enabled by default. If it is disabled by default, we're back to square one - if you know you can hit this issue with exceptions causing "instantaneous" iterations and want to guard against it, you can just use a global `try { /* your code */ } catch(e) {sleep(1); throw e; }` to avoid the issue and use [thresholds that abort the test](https://docs.k6.io/docs/thresholds#section-aborting-a-test-when-a-threshold-is-crossed) early if needed.

If it is enabled by default, writing tests can become a lot more complicated. You'd have to very carefully guard and null-check everything. Especially code that parses and uses data from the responses of the tested website, which may become unreliable under the heavy load. In a way, enabling this behavior by default could potentially cause people to have the opposite global `try{} catch(e) {}` blocks just so they suppress those errors and don't needlessly abort the script.

**Comment by MaximizedAction** [comment]
I hadn't thought of a global try-catch plus threshold as a workaround :), that's cool. Not the prettiest workaround, but it's a workaround.

Yeah, having it enabled by default is very bad imo and can lead to bad things, as you wrote, when working with response data. So a setting like that should definitely be off by default.

**Comment by na--** [comment]
Closing this issue, since one of the solutions discussed here (minimum iteration duration) was implemented in [this pull request](https://github.com/loadimpact/k6/issues/818) and was [released as part of k6 0.23.0](https://github.com/loadimpact/k6/blob/master/release%20notes/v0.23.0.md#new-option-set-a-minimum-iteration-duration-821). I also created a new, separate issue for the other idea partially mentioned above (and in a few other places): https://github.com/loadimpact/k6/issues/877
